src/charm.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charmed Kubernetes Operator for the PostgreSQL database."""
+
+import logging
+import secrets
+import string
+
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus
+from ops.pebble import Layer
+
+logger = logging.getLogger(__name__)
+
+
+class PostgresqlOperatorCharm(CharmBase):
+    """Charmed Operator for the PostgreSQL database."""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self._postgresql_service = "postgresql"
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.postgresql_pebble_ready, self._on_postgresql_pebble_ready)
+
+    def _on_install(self, _):
+        """Event handler for InstallEvent."""
+        # TODO: change to peer/leader data bag when relations are implemented.
+        self._stored.postgres_password = self._new_password()
```

Comment:
Correct me if I am wrong, but I think this can be done without relations being implemented (this is because peers refer to units from the same application as according to the [Juju SDK](https://juju.is/docs/sdk/relations#heading--peer-relations), I believe `self.peers` should still be accessible and can be used to save passwords by doing something like:

```
        root_password = None

        data = self.peers.data[self.peers.app]
        root_password = data.get('root_password', None)

        if root_password is None:
            root_password = self._new_password()
            <set peer password>
```

Reply:
This is correct - so long as you have a `peers` section in metadata :)

One thing worth noting, is that you'll want to guard writes to the `self.peers.data[self.app]` attribute with `if self.unit.is_leader()`. Similarly to provides/requires relation, units can write to their own data bag, but only leaders can write to the *application* data bag :)

Reply:
Thanks for the great contribution Mia and Jon. I used the `_on_leader_elected` event in order to enable just the leader to write the data. The changes are in the last commit.